app/scripts/lib/rpc-method-middleware/handlers/send-metadata.ts

```diff
@@ -0,0 +1,84 @@
+import { ethErrors } from 'eth-rpc-errors';
+import type { SubjectType } from '@metamask/permission-controller';
+import type {
+  Json,
+  JsonRpcParams,
+  JsonRpcRequest,
+  JsonRpcResponse,
+} from '@metamask/utils';
+import type {
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app';
+import { AddSubjectMetadata, HandlerWrapperType } from './handlers-helper';
+
+type SendMetadataOptionsType = {
+  addSubjectMetadata: AddSubjectMetadata;
+  subjectType: SubjectType;
+};
+
+type SendMetadataType<
+  Params extends JsonRpcParams = JsonRpcParams,
+  Result extends Json = Json,
+> = {
+  implementation: (
+    _req: JsonRpcRequest<Params>,
+    res: JsonRpcResponse<Result>,
+    _next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+    { addSubjectMetadata, subjectType }: SendMetadataOptionsType,
+  ) => void;
+} & HandlerWrapperType;
+
+/**
+ * This internal method is used by our external provider to send metadata about
+ * permission subjects so that we can e.g. display a proper name and icon in
+ * our UI.
+ */
+
+const sendMetadata: SendMetadataType = {
+  methodNames: [MESSAGE_TYPE.SEND_METADATA],
+  implementation: sendMetadataHandler,
+  hookNames: {
+    addSubjectMetadata: true,
+    subjectType: true,
+  },
+};
+export default sendMetadata;
+
+/**
+ * @param req - The JSON-RPC request object.
+ * @param res - The JSON-RPC response object.
+ * @param _next - The json-rpc-engine 'next' callback.
+ * @param end - The json-rpc-engine 'end' callback.
+ * @param options
+ * @param options.addSubjectMetadata - A function that records subject
+ * metadata, bound to the requesting origin.
+ * @param options.subjectType - The type of the requesting origin / subject.
+ */
+function sendMetadataHandler<Params extends JsonRpcParams, Result extends Json>(
+  req: JsonRpcRequest<Params>,
+  res: JsonRpcResponse<Result>,
```

Comment:
I'm seeing implicit `any`s for `req` and `res`, resulting in a bunch of implicit `any`s throughout the function body, and downstream objects/types as well.

Reply:
I'm seeing the same thing. `InferWithParams` in `@metamask/utils` seems to be resolving to `any` only in this branch, and not in the utils package itself. This is weird..